Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -108,63 +108,52 @@ extension HTTPClient {
                 }
             }
+        }
 
-            func socketPathFromURL(_ url: URL) throws -> String {
-                switch self {
-                case .unixSocket(.baseURL):
-                    return url.baseURL?.path ?? url.path
-                case .unixSocket:
-                    guard let socketPath = url.host else {
-                        throw HTTPClientError.missingSocketPath
-                    }
-                    return socketPath
-                case .host:
-                    return ""
-                }
-            }
+        static func useTLS(_ scheme: String) -> Bool {
+            return scheme == "https" || scheme == "https+unix"
+        }
 
-            func uriFromURL(_ url: URL) -> String {
-                switch self {
-                case .host:
-                    return url.uri
-                case .unixSocket(.baseURL):
-                    return url.baseURL != nil ? url.uri : "/"
-                case .unixSocket:
-                    return url.uri
-                }
+        static func deconstructURL(
+            _ url: URL
+        ) throws -> (
+            kind: Kind, scheme: String, hostname: String, port: Int, socketPath: String, uri: String
+        ) {
+            guard let scheme = url.scheme?.lowercased() else {
+                throw HTTPClientError.emptyScheme
             }
-
-            func supportsRedirects(to scheme: String?) -> Bool {
-                guard let scheme = scheme?.lowercased() else { return false }
-
-                switch self {
-                case .host:
-                    return Kind.hostRestrictedSchemes.contains(scheme)
-                case .unixSocket:
-                    return Kind.allSupportedSchemes.contains(scheme)
+            switch scheme {
+            case "http", "https":
+                guard let host = url.host, !host.isEmpty else {
+                    throw HTTPClientError.emptyHost
+                }
+                let defaultPort = self.useTLS(scheme) ? 443 : 80
+                return (.host, scheme, host, url.port ?? defaultPort, "", url.uri)
+            case "http+unix", "https+unix":
+                guard let socketPath = url.host, !socketPath.isEmpty else {
+                    throw HTTPClientError.missingSocketPath
+                }
+                let (kind, defaultPort) = self.useTLS(scheme) ? (Kind.UnixScheme.https_unix, 443) : (.http_unix, 80)
+                return (.unixSocket(kind), scheme, "", url.port ?? defaultPort, socketPath, url.uri)
+            case "unix":
+                let socketPath = url.baseURL?.path ?? url.path
+                let uri = url.baseURL != nil ? url.uri : "/"
+                guard !socketPath.isEmpty else {
+                    throw HTTPClientError.missingSocketPath
                 }
+                return (.unixSocket(.baseURL), scheme, "", url.port ?? 80, socketPath, uri)
+            default:
+                throw HTTPClientError.unsupportedScheme(url.scheme!)
```

Comment:
```suggestion
                throw HTTPClientError.unsupportedScheme(scheme)
```
We have a non-optional scheme already in scope, no need to force unwrap.

Reply:
The non-optional scheme is lowercased, and I figured it would preserve existing behaviour better to retain the original scheme verbatim. But I don't mind changing it if you think that isn't worth doing.

Reply:
Makes sense, let's keep it this way.